### PR TITLE
feat(elasticsearch): advanced query, identity autocomplete, exact match weight

### PR DESCRIPTION
**Pull request by david-leifker**
- Restore advanced field level queries using `/q` for backwards compatibility
- Enable autocomplete on corp users/groups
- Documentation updates around advanced queries
- Add additional weight to exact vs prefix match

## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

**Comment by david-leifker** [comment]
Need to look at that search test closer. Probably needs addressing.

**Comment by xiphl** [comment]
Thanks for updating the search guide! We specifically covered the ways to search as part of our on boarding for new users, and it will be very helpful to have examples

File: metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/query/request/SearchQueryBuilder.java [hunk 1]
```diff
@@ -89,41 +94,46 @@ private static Set<Pair<String, Float>> getStandardFields(@Nonnull EntitySpec en
...
+              if (query.contains("\"")) {
+                finalQuery.should(QueryBuilders.termQuery(fieldSpec.getFieldName() + ".keyword", query.replaceAll("\"", ""))
+                        .boost(Float.parseFloat((String) PRIMARY_URN_SEARCH_PROPERTIES.get("boostScore")) * EXACT_MATCH_BOOST_FACTOR)
+                        .queryName(fieldSpec.getFieldName() + ".keyword"));
+              }
+            });
     return finalQuery.should().size() > 0 ? Optional.of(finalQuery) : Optional.empty();
   }
 
-  private static QueryBuilder getPhraseQuery(@Nonnull EntitySpec entitySpec, String query) {
```
Comment (szalai1) [inline metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/query/request/SearchQueryBuilder.java:119]: dead code?

**Review by szalai1 — APPROVED** [review]

File: docs/how/search.md [hunk 2]
```diff
@@ -84,36 +84,47 @@ With better metadata comes better results. Learn more about ingestion technical
...
-5. Find a dataset with the term **logical** in the dataset description  
-```editedDescription: *logical* OR description: *logical*``` [Sample results](https://demo.datahubproject.io/search?page=1&query=editedDescription%3A%20%2Alogical%2A%20OR%20description%3A%20%2Alogical%2A)  
-Similar to field descriptions, dataset descriptions can be found in 2 aspects, hence the need to search 2 attributes.  
-
-6. Find a dataset which reside in one of the browsing folders, for instance, the **hive** folder  
-```browsePaths: *hive*``` [Sample results](https://demo.datahubproject.io/search?page=1&query=browsePaths%3A%20%2Ahive%2A)  
-BrowsePath is stored as a complete string, for instance ```/datasets/prod/hive/SampleKafkaDataset```, hence the need for wildcards on both ends of the term to return a result. 
+If you want to:
+
+- Exact match on term or phrase
```
Comment (jjoyce0510) [inline docs/how/search.md:92]: Thank you!!!

File: metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/query/request/SearchQueryBuilder.java [hunk 3]
```diff
@@ -32,6 +32,9 @@
 
 
 public class SearchQueryBuilder {
+
+  public static final String STRUCTURED_QUERY_PREFIX = "\\\\/q ";
+  public static final float EXACT_MATCH_BOOST_FACTOR = 10.0f;
```
Comment (jjoyce0510) [inline metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/query/request/SearchQueryBuilder.java:37]: Exact match on ANY field?

File: metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/query/request/SearchQueryBuilder.java [hunk 4]
```diff
@@ -89,41 +94,46 @@ private static Set<Pair<String, Float>> getStandardFields(@Nonnull EntitySpec en
...
 
     return fields;
   }
 
   private static Optional<QueryBuilder> getPrefixQuery(@Nonnull EntitySpec entitySpec, String query) {
-    BoolQueryBuilder finalQuery =  QueryBuilders.boolQuery();
+    BoolQueryBuilder finalQuery = QueryBuilders.boolQuery();
+    if (query.contains("\"")) {
+      finalQuery.should(QueryBuilders.termQuery("urn", query.replaceAll("\"", ""))
+              .boost(Float.parseFloat((String) PRIMARY_URN_SEARCH_PROPERTIES.get("boostScore")) * EXACT_MATCH_BOOST_FACTOR)
```
Comment (jjoyce0510) [inline metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/query/request/SearchQueryBuilder.java:115]: Wait wait so this on occurs if the query contains " " quotations (at least one) and only applies to the _urn_ field?

Is that really what we want? My experience is that exact match on _NAME_ fields is not very good, which may or may not be in the URN at all. 

Also, do users really want to have to put quotes in order to do exact match? If I search for a Table with the exact name I expect to see it - quotes or no quotes.

File: metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/query/request/SearchQueryBuilder.java [hunk 5]
```diff
@@ -89,41 +94,46 @@ private static Set<Pair<String, Float>> getStandardFields(@Nonnull EntitySpec en
...
+            .filter(SearchableAnnotation::isEnableAutocomplete) // Proxy for identifying likely exact match fields
             .filter(e -> TYPES_WITH_DELIMITED_SUBFIELD.contains(e.getFieldType()))
-            .forEach(fieldSpec -> finalQuery.should(
-                    QueryBuilders.matchPhrasePrefixQuery(fieldSpec.getFieldName() + ".delimited", query)
-                            .boost((float) fieldSpec.getBoostScore())));
+            .forEach(fieldSpec -> {
+              finalQuery.should(QueryBuilders.matchPhrasePrefixQuery(fieldSpec.getFieldName() + ".delimited", query)
+                      .boost((float) fieldSpec.getBoostScore())
+                      .queryName(fieldSpec.getFieldName())); // less than exact
+              if (query.contains("\"")) {
```
Comment (jjoyce0510) [inline metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/query/request/SearchQueryBuilder.java:128]: This seems like kinda opaque functionality.

If the user provides quotes for a field, we automatically hit the keyword index?

File: metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/query/request/SearchQueryBuilder.java [hunk 6]
```diff
@@ -89,41 +94,46 @@ private static Set<Pair<String, Float>> getStandardFields(@Nonnull EntitySpec en
...
+                finalQuery.should(QueryBuilders.termQuery(fieldSpec.getFieldName() + ".keyword", query.replaceAll("\"", ""))
+                        .boost(Float.parseFloat((String) PRIMARY_URN_SEARCH_PROPERTIES.get("boostScore")) * EXACT_MATCH_BOOST_FACTOR)
```
Comment (jjoyce0510) [inline metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/query/request/SearchQueryBuilder.java:130]: A bit confused, why are we using `PRIMARY_URN_SEARCH_PROPERTIES` here? 

Also, can we factor out the constants?

Comment (jjoyce0510) [inline metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/query/request/SearchQueryBuilder.java:130]: So again expressing what I mentioned above, boosting cases where there are quotations in the query string I don't think will be sufficient for what users expect. They expect that searching for a name of a Table by exact string will find that table, with or without quotations

File: metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/query/request/SearchRequestHandler.java [hunk 7]
```diff
@@ -247,14 +249,16 @@ private List<AggregationBuilder> getAggregations() {
...
     highlightBuilder.postTags("");
+
     // Check for each field name and any subfields
-    _defaultQueryFieldNames.forEach(fieldName -> highlightBuilder
-            .field(fieldName)
-            .field(fieldName + ".*"));
-    highlightBuilder.field("urn.delimited");
+    _defaultQueryFieldNames.stream()
+            .flatMap(fieldName -> Stream.of(fieldName, fieldName + ".*")).distinct()
+            .forEach(highlightBuilder::field);
```
Comment (jjoyce0510) [inline metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/query/request/SearchRequestHandler.java:260]: Did we verify the column name highlighting still works after this?

File: metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/query/request/SearchRequestHandler.java [hunk 8]
```diff
@@ -247,14 +249,16 @@ private List<AggregationBuilder> getAggregations() {
...
+
     // Don't set tags to get the original field value
     highlightBuilder.preTags("");
     highlightBuilder.postTags("");
+
     // Check for each field name and any subfields
-    _defaultQueryFieldNames.forEach(fieldName -> highlightBuilder
-            .field(fieldName)
-            .field(fieldName + ".*"));
-    highlightBuilder.field("urn.delimited");
```
Comment (jjoyce0510) [inline metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/query/request/SearchRequestHandler.java:257]: Is removing this safe? Is there history showing why we added this initially?

File: metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/query/request/SearchRequestHandler.java [hunk 9]
```diff
@@ -288,6 +293,12 @@ private List<MatchedField> extractMatchedFields(@Nonnull Map<String, HighlightFi
         highlightedFieldNamesAndValues.get(fieldName.get()).add(fieldValue.string());
       }
     }
+    // fallback matched query, non-analyzed field
+    for (String queryName : hit.getMatchedQueries()) {
```
Comment (jjoyce0510) [inline metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/query/request/SearchRequestHandler.java:297]: What is the purpose of this change? I'm not familiar enough

File: metadata-io/src/test/java/com/linkedin/metadata/search/elasticsearch/fixtures/SampleDataFixtureTests.java [hunk 10]
```diff
@@ -624,6 +664,241 @@ public void testFragmentUrns() {
...
+        result = search(searchService, query);
+        assertTrue(result.hasEntities() && !result.getEntities().isEmpty(),
+                String.format("%s - Expected search results", query));
+        assertTrue(result.getEntities().stream().noneMatch(e -> e.getMatchedFields().isEmpty()),
+                String.format("%s - Expected search results to include matched fields", query));
+        assertEquals(result.getEntities().size(), 2);
+    }
+
+    @Test
+    public void testPrefixVsExact() {
```
Comment (jjoyce0510) [inline metadata-io/src/test/java/com/linkedin/metadata/search/elasticsearch/fixtures/SampleDataFixtureTests.java:887]: awesome tests!

File: metadata-models/src/main/pegasus/com/linkedin/identity/CorpGroupInfo.pdl [hunk 11]
```diff
@@ -20,6 +20,7 @@ record CorpGroupInfo {
   @Searchable = {
     "fieldType": "TEXT_PARTIAL"
     "queryByDefault": true,
+    "enableAutocomplete": true,
```
Comment (jjoyce0510) [inline metadata-models/src/main/pegasus/com/linkedin/identity/CorpGroupInfo.pdl:23]: What this previously enabled and now it is removed? Curious about the reasoning for including this? Or did a customer ask for it?

File: metadata-models/src/main/pegasus/com/linkedin/identity/CorpUserInfo.pdl [hunk 12]
```diff
@@ -90,6 +91,7 @@ record CorpUserInfo includes CustomProperties {
   @Searchable = {
     "fieldType": "TEXT_PARTIAL",
     "queryByDefault": true,
+    "enableAutocomplete": true,
```
Comment (jjoyce0510) [inline metadata-models/src/main/pegasus/com/linkedin/identity/CorpUserInfo.pdl:94]: Was this removed? I recall we used to have this

File: metadata-models/src/main/pegasus/com/linkedin/metadata/key/CorpGroupKey.pdl [hunk 13]
```diff
@@ -11,7 +11,10 @@ record CorpGroupKey {
   * The URL-encoded name of the AD/LDAP group. Serves as a globally unique identifier within DataHub.
   */
   @Searchable = {
-    "fieldType": "TEXT_PARTIAL"
+    "fieldType": "TEXT_PARTIAL",
+    "queryByDefault": true,
```
Comment (jjoyce0510) [inline metadata-models/src/main/pegasus/com/linkedin/metadata/key/CorpGroupKey.pdl:15]: Nice, thanks!

File: metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/query/request/SearchQueryBuilder.java (hunk 3 above)
Comment (david-leifker) [reply metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/query/request/SearchQueryBuilder.java:37]: It's a multiplier, so the annotation * this factor.

File: metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/query/request/SearchQueryBuilder.java (hunk 4 above)
Comment (david-leifker) [reply metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/query/request/SearchQueryBuilder.java:115]: The urn field is not part of any of the models' search specs. It is always treated as an addition added everywhere. This would only be used in a quoted urn, exact match. Below is where we handle the other fields. 
More important for the non-urn field, the quotes become important when considering synonyms, stemming, and other tokenization steps should not be considered. Fulltext search is not the same as a query where exact matches are implicit, both the field and the search terms are manipulated for getting relevant results over exact match.

File: metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/query/request/SearchQueryBuilder.java (hunk 5 above)
Comment (david-leifker) [reply metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/query/request/SearchQueryBuilder.java:128]: See comment above.

File: metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/query/request/SearchQueryBuilder.java (hunk 6 above)
Comment (david-leifker) [reply metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/query/request/SearchQueryBuilder.java:130]: Err, this should. not be based on urn properties. Fixing.

Comment (david-leifker) [reply metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/query/request/SearchQueryBuilder.java:130]: I'll remove the quote conditional

File: metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/query/request/SearchRequestHandler.java (hunk 7 above)
Comment (david-leifker) [reply metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/query/request/SearchRequestHandler.java:260]: Yes

File: metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/query/request/SearchRequestHandler.java (hunk 9 above)
Comment (david-leifker) [reply metadata-io/src/main/java/com/linkedin/metadata/search/elasticsearch/query/request/SearchRequestHandler.java:297]: For exact match, keyword fields + term queries are not analyzed, therefore they do not return highlighted fields. This is to populate which field is a matching when the field is unanalyzed. Let me see if I can populate the value. The queries I've seen that hit this were only returning the urn, however if the hit does have the matched field, it could be used to fill the value in.

File: metadata-models/src/main/pegasus/com/linkedin/identity/CorpGroupInfo.pdl (hunk 11 above)
Comment (david-leifker) [reply metadata-models/src/main/pegasus/com/linkedin/identity/CorpGroupInfo.pdl:23]: You're saying this was enabled previously and removed? I think it is natural to search for the display name of the group. I added this specifically because a customer was running a search and I'd suggested autocomplete for users/groups. I looked to write a test for this (in this pr in fact) and it didn't work because of missing annotations.

File: metadata-models/src/main/pegasus/com/linkedin/identity/CorpUserInfo.pdl (hunk 12 above)
Comment (david-leifker) [reply metadata-models/src/main/pegasus/com/linkedin/identity/CorpUserInfo.pdl:94]: I didn't remove it, so makes sense to add it, especially given that I am suggesting enabling autocomplete for users/groups. Also I added tests for autocomplete on users/groups to ensure it works as expected.
